internal/pipeline/maybe_scalar_steps.go

```diff
@@ -0,0 +1,115 @@
+package pipeline
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/buildkite/agent/v3/internal/ordered"
+	"github.com/buildkite/interpolate"
+)
+
+// In the buildkite pipeline yaml, some step types (broadly, wait steps, input steps and block steps) can be represented
+// either by a scalar string (ie "wait") or by a mapping with keys and values and such
+//
+// This behaviour is difficult to cleanly model in go, which leads to the somewhat odd structure of the structs defined
+// in this file - each type (WaitStep, InputStep) has a Scalar field which is set to the scalar value if the step was
+// if, during pipeline parsing, the step was represented as a scalar, and is left empty if the step was represented as
+// a mapping. In essence, if one of the fields of these structs is filled, the other should be empty.
+//
+// In reading this file, you may have noticed that I mentioned that there are three types of step that can be represented
+// as a scalar, but there are only two structs defined here. This is because the third type, block steps, are represented
+// in exactly the same way as input steps, so they can share the same struct. This is liable to change in the future,
+// as conceptually they're different types, and it makes sense to have them as different types in go as well.
+//
+// Also also! The implementations for WaitStep and InputStep **almost**, but not quite identical. This is due to the behaviour
+// of marshalling an empty struct into into JSON. For WaitStep, it makes sense that the empty &WaitStep{} struct marshals
+// to "wait", but with InputStep, there's no way to tell whether it should be marshalled to "input" or "block", which
+// have very different behaviour on the backend.
+
+var validStepScalars = []string{"wait", "waiter", "block", "input", "manual"}
+
+func NewScalarStep(s string) (Step, error) {
+	switch s {
+	case "wait", "waiter":
+		return &WaitStep{Scalar: s}, nil
+	case "block", "input", "manual":
+		return &InputStep{Scalar: s}, nil
+	default:
+		return nil, fmt.Errorf("unmarshaling step: unsupported scalar step type %q, want one of %v", s, validStepScalars)
+	}
+}
+
+// WaitStep models a wait step.
+//
+// Standard caveats apply - see the package comment.
+type WaitStep struct {
+	Scalar   string         `yaml:"-"`
+	Contents map[string]any `yaml:",inline"`
+}
+
+// MarshalJSON marshals a wait step as "wait" if w is empty, or as the step's scalar if it's set.
+// If scalar is empty, it marshals as the remaining fields
+func (s *WaitStep) MarshalJSON() ([]byte, error) {
+	if s.Scalar == "" {
+		if len(s.Contents) == 0 {
+			return json.Marshal("wait")
+		}
+		return json.Marshal(s.Contents)
+	}
+
+	return json.Marshal(s.Scalar)
```

Comment:
Very low-priority hobby-horse of mine: 

```suggestion
	if s.Scalar != "" {
	  return json.Marshal(s.Scalar)
	}
	if len(s.Contents) == 0 {
		return json.Marshal("wait")
	}
	return json.Marshal(s.Contents)
```

Reply:
🙈 i can't believe i missed this